### PR TITLE
[Bug-3187]close Heartbeat thread pool when MasterRegistry unRegistry

### LINKED ISSUE #3187: [BUG] Heartbeat thread pool does not shutdown when MasterRegistry unRegistry
*For better global communication, please give priority to using English description, thx! *

**Describe the bug**
look at the method of MasterRegistry
```java
    public void unRegistry() {
        String address = getLocalAddress();
        String localNodePath = getMasterPath();
        zookeeperRegistryCenter.getZookeeperCachedOperator().remove(localNodePath);
        logger.info("master node : {} unRegistry to ZK.", address);
    }
```
The method, which is invoke when close the MasterServer, does not shutdown the Heartbeat thread pool.

### PR DESCRIPTION
**Pull request by tswstarplanet**
fix #3187 
## *Tips*
- *Thanks very much for contributing to Apache DolphinScheduler.*
- *Please review https://dolphinscheduler.apache.org/en-us/community/index.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds checkstyle plugin.)*

## Brief change log

Close Heartbeat thread pool when MasterRegistry unRegistry.
Add unit test

## Verify this pull request

*(Please pick either of the following options)*

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added dolphinscheduler-dao tests for end-to-end.*
  - *Added CronUtilsTest to verify the change.*
  - *Manually verified the change by testing locally.*

### CODE REVIEW

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java [hunk 1]
```diff
@@ -137,4 +142,7 @@ private String getLocalAddress(){
 
     }
 
+    public ScheduledExecutorService getHeartBeatExecutor() {
+        return heartBeatExecutor;
+    }
```
Comment (CalvinKirs) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java:147]: It’s a bit redundant here. If it’s just for testing, I don’t think it’s necessary.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java [hunk 2]
```diff
@@ -113,6 +110,14 @@ public void stateChanged(CuratorFramework client, ConnectionState newState) {
...
         String address = getLocalAddress();
         String localNodePath = getMasterPath();
+        heartBeatExecutor.shutdown();
+        try {
+            if (heartBeatExecutor.awaitTermination(masterConfig.getMasterHeartbeatInterval(), TimeUnit.SECONDS)) {
+                logger.warn("The heartbeat executor does not shutdown");
+            }
+        } catch (InterruptedException e) {
+            logger.warn("The heartbeat executor is interrupted");
+        }
```
Comment (yangyichao-mango) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java:120]: Hi,

Good job, is it better to refer to WorkerRegistry[1] to use `heartBeatExecutor.shutdownNow()`?

[1] https://github.com/apache/incubator-dolphinscheduler/blob/dev/dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/registry/WorkerRegistry.java

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java (hunk 1 above)
Comment (yangyichao-mango) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java:147]: > It’s a bit redundant here. If it’s just for testing, I don’t think it’s necessary.

I agree with @CalvinKirs .

**Comment by yangyichao-mango** [comment]
Hi,

Good job, LGTM.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java (hunk 2 above)
Comment (tswstarplanet) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistry.java]: solved

**Pull request merged**